### PR TITLE
[CHORE] Lint enforcement of type-only imports

### LINKED ISSUE #6804: We should lint against type only imports using es style
If an import is only used for type information, typescript offers us a mechanism to ensure we only import the type. Using this style both prevents cycles in rollup and helps to ensure we have not broken package encapsulation.

We should find or write a lint rule such that an import using ES syntax being used only for type information is disallowed.

e.g.

Not

```ts
import Snapshot from './snapshot';

function withSnapshot(snapshot: Snapshot) {}
```

but

```ts
type Snapshot = import('./snapshot').default;

function withSnapshot(snapshot: Snapshot) {}
```

### PR DESCRIPTION
**Pull request by runspired**
resolves #6804 

Also runs the `--fix` for it.

~~This likely will reduce our asset-sizes.~~
**Edit:** looks like thanks to previous manual optimization of this + rollup we don't see a win here (previous manual optimization did result in wins). But this will help avoid future regressions :) 

Will revisit #6849 once this lands.

### CODE REVIEW

**Comment by igorT** [comment]
This lint rule seems more generically useful, should we publish it as a separate repo so other people can use it?

File: packages/unpublished-eslint-rules/index.js [hunk 1]
```diff
@@ -0,0 +1,288 @@
+const TypeParamNamingSchema = [['T'], ['K', 'V'], ['T', 'K', 'V'], ['T', 'K', 'V', 'R']];
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: Is the structure of this code copied and adapter from somewhere or is this a 100% greenfield effort?

**Comment by igorT** [comment]
Some things that you import as `type` have been `interfaces` while others have been classes, are there any consequences to this intermingling?

File: packages/unpublished-eslint-rules/index.js (hunk 1 above)
Comment (runspired) [reply packages/unpublished-eslint-rules/index.js]: It's greenfield

File: packages/unpublished-eslint-rules/index.js [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+              local: node.local,
+              imported: node.imported || node.local,
+              from: node.parent.source,
+              isDefault: false,
+            };
+          },
+          Identifier(node) {
+            let type = node.parent.type;
+
+            if (type === 'ImportSpecifier' || type === 'ImportDefaultSpecifier') {
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: We can probably ignore more types here, like declarations etc.

File: packages/unpublished-eslint-rules/index.js [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+              // console.log(`found ${node.name} for import type ${type}`);
+            } else if (type === 'TSTypeReference') {
+              // console.log(`found type reference ${node.name}`);
+              // do nothing, this Identifier is handled elsewhere
+            } else if (imports[node.name] !== undefined) {
+              const scope = context.getScope();
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: const?

File: packages/unpublished-eslint-rules/index.js [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+              if (scope.type === 'class') {
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: Lets make the top if a switch or something like that, to make these nested ifs easier to read

File: packages/unpublished-eslint-rules/index.js [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            usedTypes = {};
+            lastImport = null;
+          },
+          'Program:exit'() {
+            // console.log('\n\n\nProgram:exitBegin\n======\n');
+            // eslint will only pass over a file 10x before bailing
+            // our approach lets us autofix more things by collapsing fixes for
+            // a single import declaration into one error+change instead of fix
+            // by individual imports within a declaration
+            for (let i = 0; i < declarations.length; i++) {
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: Maybe use something like forEach etc.

File: packages/unpublished-eslint-rules/index.js [hunk 6]
```diff
@@ -0,0 +1,285 @@
+const TypeParamNamingSchema = [['T'], ['K', 'V'], ['T', 'K', 'V'], ['T', 'K', 'V', 'R']];
+const Alphabet = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P'];
+
+function getUnusedSpecifiers(declaration, used) {
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: Can we write a comment for this

File: packages/unpublished-eslint-rules/index.js [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        let imports, usedImports, usedTypes, lastImport, declarations;
+        return {
+          Program() {
+            declarations = [];
+            imports = {};
+            usedImports = {};
+            usedTypes = {};
+            lastImport = null;
+          },
+          'Program:exit'() {
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: This method seems kinda large, maybe would be nice to break up a bit?

File: packages/unpublished-eslint-rules/index.js [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+            // console.log('\n\n\nProgram:exitBegin\n======\n');
+            // eslint will only pass over a file 10x before bailing
+            // our approach lets us autofix more things by collapsing fixes for
+            // a single import declaration into one error+change instead of fix
+            // by individual imports within a declaration
+            for (let i = 0; i < declarations.length; i++) {
+              let declaration = declarations[i];
+              let unusedImports = getUnusedSpecifiers(declaration, usedImports);
+              let typeImports = getTypeSpecifiers(unusedImports, usedTypes);
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: Maybe rename this to typeOnlyImports?

File: packages/unpublished-eslint-rules/index.js [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+          },
+          TSTypeReference(node) {
+            let identifier = node.typeName;
+            if (imports[identifier.name] !== undefined) {
+              // console.log(`found used type ${identifier.name}`);
+              if (node.typeParameters && node.typeParameters.params.length) {
+                // this type usage contains generics e.g. `Dict<Thing>`
+                // so we save off how many params it uses to fix up the import
+                // later
+                usedTypes[identifier.name] = node.typeParameters.params.length;
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: Maybe create another object to keep the type params so we don't have both booleans and numbers in this one

File: packages/unpublished-eslint-rules/index.js [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+    }
+    for (let i = 0; i < paramValue; i++) {
+      params.push(alphabet[i]);
+    }
+    paramString = `<${params.join(', ')}>`;
+  }
+  return paramString;
+}
+
+function getRangeToRemove(exportToRemove) {
```
Comment (igorT) [inline packages/unpublished-eslint-rules/index.js]: Can we describe this method a bit more please

**Pull request merged**